test/util/framework/per_test_framework.go

```diff
@@ -0,0 +1,273 @@
+// Copyright 2025 Microsoft Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package framework
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/resources/armresources"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/resources/armsubscriptions"
+	"github.com/onsi/ginkgo/v2"
+	"golang.org/x/sync/errgroup"
+	"k8s.io/apimachinery/pkg/util/rand"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+
+	hcpapi20240610 "github.com/Azure/ARO-HCP/internal/api/v20240610preview/generated"
+)
+
+type perItOrDescribeTestContext struct {
+	perBinaryInvocationTestContext *perBinaryInvocationTestContext
+
+	contextLock                   sync.RWMutex
+	knownResourceGroups           []string
+	subscriptionID                string
+	clientFactory20240610         *hcpapi20240610.ClientFactory
+	armResourcesClientFactory     *armresources.ClientFactory
+	armSubscriptionsClientFactory *armsubscriptions.ClientFactory
+}
+
+func NewInvocationContext() *perItOrDescribeTestContext {
+	tc := &perItOrDescribeTestContext{
+		perBinaryInvocationTestContext: invocationContext(),
+	}
+
+	// this construct allows us to be called before or after the test has started and still registry proper cleanup.
+	if !ginkgo.GetSuite().InRunPhase() {
+		// The order is important here: if the extension calls ginkgo.BeforeEach
```

Comment:
Ginkgo is some kind of brain-melting nonsense - for those of us not steeped in this bizzaro-world (and unhappy that this thing just won't die), can you better explain what an extension is, what the callback is, etc?

Reply:
copied  comment.  There is an ordering, but it doesn't matter here.